### PR TITLE
[Nano] Nano How-to Guides: Format & PyTorch Inference

### PR DESCRIPTION
**Pull request by Oscilloscope98**
### 1. Why the change?

Add Nano how-to guides to our documentation for better user experience. How-to guides aim at giving multiple bite-sized, task-oriented, and executable examples for user to check when they need to do similar tasks.

In this PR, guides for PyTorch inference is covered. Meanwhile, this PR also defines the format/workflow of creating Nano How-To guides.

### 2. Summary of the change
The guides source file are located under folder `BigDL/python/nano/tutorial/notebook`. We link this folder to sphinx root folder and render the notebook to the web page directly.

The notebook looks good in both local notebook viewer and github preview.

- [x] How-to Guide Index Page (draft)

#### Inference
PyTorch
- [x]  How to accelerate a PyTorch inference pipeline through ONNXRuntime (based on [this](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/inference/pytorch/pytorch_inference_onnx.py))
- [x]  How to accelerate a PyTorch inference pipeline through OpenVINO (based on [this](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/inference/pytorch/pytorch_inference_openvino.py))
- [x]  How to quantize your PyTorch model for inference using Intel Neural Compressor (based on [this](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/inference/pytorch/pytorch_quantization.py) without runtime acceleration, and [this](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/inference/pytorch/pytorch_quantization_onnx.py) with runtime acceleration)
- [x]  How to quantize your PyTorch model for inference using Post-training Optimization Tools (based on [this](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/inference/pytorch/pytorch_quantization_openvino.py#L33))

#### General
- [x] How to run Nano on Colab (wip, need to be updated when more problems met)

### 3. How to test?
- [x] Document test: https://yuwentestdocs.readthedocs.io/en/nano-how-to-guides/doc/Nano/Howto/index.html
- [ ] Notebook test on github action (will be done in further prs)

Notebook test locally (conda create an empty environment with python=3.7)
- [x]  How to accelerate a PyTorch inference pipeline through ONNXRuntime 
- [ ]  How to accelerate a PyTorch inference pipeline through OpenVINO (not able to do local test because the latest version of openvino-dev is not supported in CentOS)
- [x]  How to quantize your PyTorch model for inference using Intel Neural Compressor 
- [ ]  How to quantize your PyTorch model for inference using Post-training Optimization Tools (not able to do local test because the latest version of openvino-dev is not supported in CentOS)

### CODE REVIEW

**Comment by rnwang04** [comment]
1. To accelerate your PyTorch inference pipeline through ONNXRuntime as backend engine, BigDL-Nano provide you with Trainer.trace(accelerator='onnxruntime') API
2. and each pixel is associated with 3 floating-point numbers in , which represent rgb values. this line may be deleted.
3. NOTE: input_sample is the parameter for ONNXRuntime accelerator to know the shape of model input. So both the batch size and the specific values are not important to input_sample. This is the reason why we could use torch.rand(1, 3, 224, 224) for input_sample here.

**Comment by rnwang04** [comment]
```
# for running on Colab
!pip uninstall -y torchtext # uninstall torchtext to avoid version conflict

# for BigDL-Nano
!pip install --pre --upgrade bigdl-nano[pytorch] # install the nightly-bulit version
!source bigdl-nano-init # set environment variables for acceleration based on your current hardware

# for ONNXRuntime accelerator
!pip install onnx onnxruntime
!pip install --upgrade numpy
```

**Comment by Oscilloscope98** [comment]
> `!pip install --upgrade numpy` may be added to openvino.

When installing `openvino-dev`, `numpy` is downgraded from a higher version>1.20.0 to a lower version<1.20.0, due to the reason mentioned here: https://stackoverflow.com/a/69528768, error may occur when using the `Trainer.trace` function:
image

Based on Ruonan's comment, I manually update `numpy` now to solve this error
image
 
Will there be a better solution?

**Comment by shane-huang** [comment]
How-to-guides needs to be small enough to just explain one small feature, So 
* save/load, install (colab) can be separate how-to-guides and you can use reference links to those guides. 
* As we already has fundamental .py examples, we don't need to jupyter notebooks -> md.  Just make these guides md files and use code snippets from corresponding fundamental example to explain the feature. 
* Add reference links to the runnable fundamental .py example in the how-to-guide.

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Accelerate PyTorch Inference using ONNXRuntime
+
+To accelerate your PyTorch inference pipeline through ONNXRuntime as backend engine, BigDL-Nano provides you with `Trainer.trace(accelerator='onnxruntime')` API. Based on that, the optimization could be implemented by applying very few lines of code changes.
+
+Let us suppose we have a [ResNet-18 model](https://pytorch.org/vision/main/models/generated/torchvision.models.resnet18.html) pretrained on ImageNet dataset, and would like to do the inference on a randomly generated test dataset. The model would be like:
```
Comment (shane-huang) [inline docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md:5]: Let's take a [ResNet-18 model](https://pytorch.org/vision/main/models/generated/torchvision.models.resnet18.html) pretrained on ImageNet dataset as an example. First we load the model:
```python
from torchvision.models import resnet18
model_ft = resnet18(pretrained=True)
```
Then we set it to the evaluation mode. 
```python
model_ft.eval()
```

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Accelerate PyTorch Inference using ONNXRuntime
+
+To accelerate your PyTorch inference pipeline through ONNXRuntime as backend engine, BigDL-Nano provides you with `Trainer.trace(accelerator='onnxruntime')` API. Based on that, the optimization could be implemented by applying very few lines of code changes.
```
Comment (shane-huang) [inline docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md:3]: You can use `Trainer.trace(accelerator='onnxruntime')` API to enable the ONNXRuntime acceleration for pytorch inference.  It only takes a few lines.

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+```python
+x = torch.rand(2, 3, 224, 224)
+# use the optimized model here
+y_hat = ort_model(x)
+predictions = y_hat.argmax(dim=1)
+print(predictions)
+```
+
+An executable version of the example above could be found [here](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/inference/pytorch/pytorch_inference_onnx.py).
```
Comment (shane-huang) [inline docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md:41]: A short runnable example to demonstrate this functionality can be found here.

**Comment by jason-dai** [comment]
> > * As we already have fundamental .py examples, we don't need jupyter notebooks.  Just simply make these guides .md files and use code snippets from corresponding fundamental example to explain the feature.
> 
> How do we keep the guide and the py files in sync?

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md (hunk 1 above)
Comment (Oscilloscope98) [reply docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md]: Have revised :D

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md (hunk 2 above)
Comment (Oscilloscope98) [reply docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md]: Have revised

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md (hunk 3 above)
Comment (Oscilloscope98) [reply docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md]: Have revised

**Comment by shane-huang** [comment]
> > * As we already have fundamental .py examples, we don't need jupyter notebooks.  Just simply make these guides .md files and use code snippets from corresponding fundamental example to explain the feature.
> 
> How do we keep the guide the py files in sync?

I have thought about the syncing. There're several options: 
- Option1:  render a full jupyter notebook to corresponding .md.  But the full notebook may include extra steps that are not related to the specific feature, e.g. installation, which can appear in many notebooks and is redundant. Solution maybe to find a way to render only part of the notebook. 
- Option 2: embed part of example from .py file and then add explanations to it. But if someone changed the .py file but did not change how-to guides, the lines will mismatch. Solution may be to find a way to embed the code not by lines but by special semantics, e.g. some special characters in comment, etc., and we will need a unit test to ensure the quoted code is correct.
- Option 3: we make how-to-guides standalone to explain the feature, and direct user to runnable .py code for trials. 

IMHO,  I don't think syncing is a big issue. The main purpose of writing these fundamental examples/how-to-guides is to tell user how to use a specific function. The guide should be simple and straightforward. And for impatient users, he does not need to run the example to know how to solve the problem (he can run it if he wants so we attach the link to .py).

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+# use the optimized model here
+y_hat = ort_model(x)
+predictions = y_hat.argmax(dim=1)
+print(predictions)
+```
+
+A short runnable example to demonstrate this functionality can be found [here](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/inference/pytorch/pytorch_inference_onnx.py).
+
+```eval_rst
+.. card:: Relative Readings
```
Comment (shane-huang) [inline docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md:50]: Related Readings

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+```python
+from torchvision.models import resnet18
+
+model_ft = resnet18(pretrained=True)
+```
+
+Then we set it in evaluation mode:
+
+```python
+model_ft.eval()
```
Comment (zhentaocc) [inline docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md:16]: shall we move this step to `Trainer.trace`? @TheaperDeng

File: docs/readthedocs/source/doc/Nano/Howto/quantize_pytorch_inference_inc.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+model = resnet18(pretrained=True)
+_, train_dataset, val_dataset = finetune_pet_dataset(model)
+```
+> _The full definition of function_ `finetune_pet_dataset` _could be found_ [_here_](https://github.com/intel-analytics/BigDL/blob/1122517e0f919d7880e8f9848a476ce889ed7183/python/nano/tutorial/inference/pytorch/pytorch_quantization.py#L33).
+
+Then we set it in evaluation mode:
+
+```python
+model.eval()
+```
```
Comment (zhentaocc) [inline docs/readthedocs/source/doc/Nano/Howto/quantize_pytorch_inference_inc.md:19]: same here.

Comment (zhentaocc) [reply docs/readthedocs/source/doc/Nano/Howto/quantize_pytorch_inference_inc.md:19]: shall we move this step to `Trainer.trace`? @TheaperDeng

File: docs/readthedocs/source/doc/Nano/Howto/quantize_pytorch_inference_inc.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                from bigdl.nano.pytorch import Trainer
+
+                q_model = Trainer.quantize(model,
+                                           accelerator='onnxruntime',
+                                           calib_dataloader=DataLoader(train_dataset, batch_size=32))
+```
+
+```eval_rst
+.. note::
+    ``Trainer`` will by default quantize your PyTorch ``nn.Module`` through **static** post-training quantization. For this case, ``calib_dataloader`` (for calibration data) is required. Batch size is not important to ``calib_dataloader``, as it intends to read 100 data. And there could be no label in calibration data.
```
Comment (zhentaocc) [inline docs/readthedocs/source/doc/Nano/Howto/quantize_pytorch_inference_inc.md:47]: `100 samples` instead of `100 data`

**Review by zhentaocc — COMMENTED** [review]
LGTM for the rest of parts.

**Comment by Oscilloscope98** [comment]
> * Option 2: embed part of example from .py file and then add explanations to it. But if someone changed the .py file yet did not change how-to guides, the lines will mismatch. Solution may be to find a way to embed the code not by lines but by special semantics, e.g. some special characters in comment, etc., and we will need a unit test to ensure the quoted code is correct.

For Option 2, I currently find a way to realize it.

With the help of directive [`literalinclude`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude), we could add some specific and meaningful comments in the py file as marks, such as:
`# >>> import a pretrained ResNet-18 model >>>`
...
`# <<< import a pretrained ResNet-18 model <<<`
to directly include codes between these marks into the documentation.

I have made a test commit [8f79fc4](https://github.com/intel-analytics/BigDL/pull/5480/commits/8f79fc4290a647333d4e518cc9ec8422fae06721) for it. The resulted [web page](https://yuwentestdocs.readthedocs.io/en/nano-how-to-guides/doc/Nano/Howto/accelerate_pytorch_inference_onnx.html) has no change, but several comments has been added to the corresponding [py file](https://github.com/intel-analytics/BigDL/pull/5480/files#diff-d2b2ea42be2a12d0032bb6c715d3c5d63a9b7fe2cd5f2ade246bf198747c85b2).

But this kind of realization has two problems:
1. When writing the how-to guides, we may put several lines of code that are not adjacent in the py files together in the guide. For example, these two lines of codes are not adjacent in the py file.
image
I have not found a elegant way yet to quote non-adjacent codes, and put them together in a code block in the documentation.
To solve this, I currently make some tiny change to the py file directly.

2. When directly quoting the codes in the py file, there could be unwanted indentation. For example, if we quote inside a function, there will always be a tab in front of each line. I have not found a elegant way yet to solve this either.
I currently use the `dedent` option of `literalinclude` to remove the first 4 letter of each line in the sphinx project. However, I don't think this is desired, as when changing tab in py file, mismatch will be caused.

@shane-huang What do you think of this way to realize synchronization?

**Comment by Oscilloscope98** [comment]
> * Option1:  render a full jupyter notebook to corresponding .md.  But the full notebook may include extra steps that are not related to the specific feature, e.g. installation, which can appear in many notebooks and is redundant. Solution maybe to find a way to render only part of the notebook.

For Option 1, I have also found a way to realize it. We could add `'nbsphnix': 'hidden'` attribute in the `metadata` of the notebook cell we want to hide from rendering in documentation.

I have made another test commit for it

File: docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md (hunk 5 above)
Comment (jason-dai) [reply docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx.md:16]: What if the user does not set `eval`? I think we should automatically set it to `eval` in `trace` and then restore the status.

**Comment by Oscilloscope98** [comment]
> * Option 1:  render a full jupyter notebook to corresponding .md.  But the full notebook may include extra steps that are not related to the specific feature, e.g. installation, which can appear in many notebooks and is redundant. Solution maybe to find a way to render only part of the notebook.
> * Option 2: embed part of example from .py file and then add explanations to it. But if someone changed the .py file yet did not change how-to guides, the lines will mismatch. Solution may be to find a way to embed the code not by lines but by special semantics, e.g. some special characters in comment, etc., and we will need a unit test to ensure the quoted code is correct.
> * Option 3: we make how-to-guides standalone to explain the feature, and direct user to runnable .py code for trials.

|   | solution | pros | cons | notes |
| :--- | :--- | :--- |  :--- |  :--- |
|1 | use `'nbsphnix': 'hidden'` attribute to hide redundant cells | - synchronization <br> - easy for further update because we only have one file for both runnable examples and the how-to guide  | - hard to add complex web components (e.g. cards, tabs) <br> - need to mark sure some cells looks both good in docs and Colab | - need to add external links to make them still functional on Colab
| 2 | add marks in py file and quote codes inside marks directly through [`literalinclude`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude) |  - synchronization <br> - easy to add different web components (e.g. cards, tabs) | - hard to demonstrate code in different order <br> - unwanted indentation | - need to add meaningful comments as marks in py files <br> - may need to change order of codes to fit the explanation logic in documentation|
| 3 | copy related codes into documentation | - easy to add different web components (e.g. cards, tabs) <br> - Highest flexibility to show wanted codes in docs | - asynchronization | - need to add link to direct readers to runnable py files |

**Comment by shane-huang** [comment]
> > * Option 1:  render a full jupyter notebook to corresponding .md.  But the full notebook may include extra steps that are not related to the specific feature, e.g. installation, which can appear in many notebooks and is redundant. Solution maybe to find a way to render only part of the notebook.
> > * Option 2: embed part of example from .py file and then add explanations to it. But if someone changed the .py file yet did not change how-to guides, the lines will mismatch. Solution may be to find a way to embed the code not by lines but by special semantics, e.g. some special characters in comment, etc., and we will need a unit test to ensure the quoted code is correct.
> > * Option 3: we make how-to-guides standalone to explain the feature, and direct user to runnable .py code for trials.
> 
> solution	pros	cons	notes
> 1	use `'nbsphnix': 'hidden'` attribute to hide redundant cells, and render the notebook directly
> 
> test [web page](https://yuwentestdocs.readthedocs.io/en/nano-how-to-guides/doc/Nano/Howto/accelerate_pytorch_inference_onnx_sync_notebook.html)
> test [Colab notebook](https://colab.research.google.com/github/Oscilloscope98/BigDL/blob/nano-how-to-guides/docs/readthedocs/source/doc/Nano/Howto/accelerate_pytorch_inference_onnx_sync_notebook.ipynb)	- synchronization
> - easy for further update because we only have one file for both runnable examples and the how-to guide	- hard to add complex web components (e.g. cards, tabs)
> - need to make sure some cells looks both good in docs and Colab	- need to add external links to make them still functional on Colab
> 2	add marks in py file and quote codes between marks directly into documentation through [`literalinclude`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude)
> 
> test [web page](https://yuwentestdocs.readthedocs.io/en/nano-how-to-guides/doc/Nano/Howto/accelerate_pytorch_inference_onnx_sync_mark.html)
> test [py file](https://github.com/intel-analytics/BigDL/pull/5480/commits/24b87a3671555ca0f705a4b5b3a60f3d345029e5#diff-d2b2ea42be2a12d0032bb6c715d3c5d63a9b7fe2cd5f2ade246bf198747c85b2) with comments as marks	- synchronization
> - easy to add different web components (e.g. cards, tabs)	- hard to demonstrate code in different order
> - unwanted indentation	- need to add meaningful comments as marks in py files
> - may need to change order of codes to fit the explanation logic in documentation
> 3	copy related codes into documentation	- easy to add different web components (e.g. cards, tabs)
> - highest flexibility to show wanted codes in docs	- asynchronization	- need to add link to direct readers to runnable py files
> @shane-huang I made a simple table to summarize my current solution for these options and their pros and cons. Please let me know your opinion on how should we realize the synchronization. Thank you!

I think the jupyter option looks better.

**Comment by TheaperDeng** [comment]
- Accelerate PyTorch Inference using ONNXRuntime

1. remove
  
  > `input_sample` is not required if you have used an instance of `Trainer` to fit your model before, or the model is a `LightningModule` with any dataloader attached.
  
  Since we don't support input_sample to be none in future. (currently, only users with old version pytorch-lightning support this)
  
2. give a little introduction (1-2 sentences) to onnxruntime

**Comment by TheaperDeng** [comment]
- Accelerate PyTorch Inference using OpenVINO

1. remove
  
  > `input_sample` is not required if you have used an instance of `Trainer` to fit your model before, or the model is a `LightningModule` with any dataloader attached.
  
  Since we don't support input_sample to be none in future. (currently, only users with old version pytorch-lightning support this)
  
2. give a little introduction (1-2 sentences) to openvino
3. I don't think `!source bigdl-nano-init` in jupyter notebook will take effective, and accuraclly we don't recommend our users to use these env variables during the inferencing (especially for inference engines). We may just delete this line.

File: docs/readthedocs/source/doc/Nano/Howto/quantize_pytorch_inference_inc.md (hunk 6 above)
Comment (TheaperDeng) [reply docs/readthedocs/source/doc/Nano/Howto/quantize_pytorch_inference_inc.md]: yeah, we need to do this.

**Comment by TheaperDeng** [comment]
- Quantize PyTorch Model for Inference using Post-training Optimization Tools
  

1. change title to "Quantize PyTorch Model for Inference using OpenVINO Post-training Optimization Tools"
  
2. You can apply `Trainer.quantize(accelerator='openvino')` API -> You can call `Trainer.quantize` with `accelerator='openvino'`
  
3. No `!source bigdl-nano-init`
  
4. Trainer.quantize(accelerator='openvino') -> Trainer.quantize(..., accelerator='openvino')
  
5. OxfordIIITPet dataset is linked to a wrong url

**Comment by Oscilloscope98** [comment]
@TheaperDeng Have updated based on your comments. And the PR has been rebased.

Introduction to ONNXRuntion and OpenVINO is not added, as I think we could suppose readers/users have known what these two libs are when they would like to check the corresponding how-to guides.

Please have a further review. Thank you very much!! :D

File: docs/readthedocs/source/doc/Nano/Howto/install_in_colab.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+            !pip install --pre --upgrade bigdl-nano[pytorch]
+```
+
+```eval_rst
+.. warning::
+    For Google Colab hosted runtime, ``source bigdl-nano-init`` is hardly to take effect as environment variables need to be set before jupyter kernel is started.
+```
+
+To avoid version conflicts caused by `trochtext`, you should uninstall it:
```
Comment (TheaperDeng) [inline docs/readthedocs/source/doc/Nano/Howto/install_in_colab.md:36]: `torchtext`

**Review by TheaperDeng — COMMENTED** [review]
others LGTM.

**Review by TheaperDeng — APPROVED** [review]
LGTM

**Pull request merged**